lib/OpModel/TTNN/TTNNOpModelLib.cpp

```diff
@@ -152,29 +217,372 @@ getMemoryConfig(const mlir::tt::ttnn::TTNNLayoutAttr &layout) {
+            ::llvm::ArrayRef<int64_t>, ::mlir::tt::ttnn::TTNNLayoutAttr> &)>>
+    g_constraints_cache_relu_op(
+        "ReluOpInterface",
+        [](const std::tuple<
+            ::llvm::ArrayRef<int64_t>, ::mlir::tt::ttnn::TTNNLayoutAttr,
+            ::llvm::ArrayRef<int64_t>, ::mlir::tt::ttnn::TTNNLayoutAttr> &key) {
+          // open device / get existing device
+          Device *device = SingletonDeviceContext::getInstance().getDevice();
+
+          // prepare io specs
+          const ::ttnn::TensorSpec input_spec =
+              detail::getTensorSpec(std::get<0>(key), std::get<1>(key));
+          const ::ttnn::TensorSpec output_spec =
+              detail::getTensorSpec(std::get<2>(key), std::get<3>(key));
+
+          // run op constraint query
+          return ::ttnn::graph::query_op_constraints(
+              ::ttnn::relu, device, input_spec,
+              output_spec.tensor_layout().get_memory_config());
+        });
+#endif // TTMLIR_ENABLE_OPMODEL
+
 bool ReluOpInterface::isLegal(
+    const ::llvm::ArrayRef<int64_t> &inputShape,
     const mlir::tt::ttnn::TTNNLayoutAttr &inputLayout,
+    const ::llvm::ArrayRef<int64_t> &outputShape,
     const mlir::tt::ttnn::TTNNLayoutAttr &outputLayout) {
-
 #ifdef TTMLIR_ENABLE_OPMODEL
-  return true; // to wire into tt-metal with the next uplift
+  return wrapper::isLegal(g_constraints_cache_relu_op, inputShape, inputLayout,
+                          outputShape, outputLayout);
 #else
   return true;
 #endif // TTMLIR_ENABLE_OPMODEL
 }
 
 std::tuple<size_t, size_t, size_t> ReluOpInterface::getOpL1Usage(
+    const ::llvm::ArrayRef<int64_t> &inputShape,
     const mlir::tt::ttnn::TTNNLayoutAttr &inputLayout,
+    const ::llvm::ArrayRef<int64_t> &outputShape,
     const mlir::tt::ttnn::TTNNLayoutAttr &outputLayout) {
 #ifdef TTMLIR_ENABLE_OPMODEL
-  return std::make_tuple(0, 0, 0); // to wire into tt-metal with the next uplift
+  return wrapper::getL1Usage(g_constraints_cache_relu_op, inputShape,
+                             inputLayout, outputShape, outputLayout);
+#else
+  return std::make_tuple(0, 0, 0);
+#endif // TTMLIR_ENABLE_OPMODEL
+}
+
+//===----------------------------------------------------------------------===//
+// AddOp
+//===----------------------------------------------------------------------===//
+
+#ifdef TTMLIR_ENABLE_OPMODEL
+TupleCache<
+    std::tuple<::llvm::ArrayRef<int64_t>, ::mlir::tt::ttnn::TTNNLayoutAttr,
+               ::llvm::ArrayRef<int64_t>, ::mlir::tt::ttnn::TTNNLayoutAttr,
+               ::llvm::ArrayRef<int64_t>, ::mlir::tt::ttnn::TTNNLayoutAttr>,
+    ::ttnn::graph::QueryResponse,
+    std::function<::ttnn::graph::QueryResponse(
+        const std::tuple<
+            ::llvm::ArrayRef<int64_t>, ::mlir::tt::ttnn::TTNNLayoutAttr,
+            ::llvm::ArrayRef<int64_t>, ::mlir::tt::ttnn::TTNNLayoutAttr,
+            ::llvm::ArrayRef<int64_t>, ::mlir::tt::ttnn::TTNNLayoutAttr> &)>>
+    g_constraints_cache_add_op(
```

Comment:
Why did you implement caching in the first iteration? It's unnecessary complication. Is there any resource governance mechanism for cache, it can just grow out of control?

Reply:
I also think it should not be global cache, it should be owned and managed by compiler component using it(Optimizer/Corresponding analysis).

Reply:
I'm fine to remove it.